xfree/xc/programs/xedit/xedit.man
.\" $XConsortium: xedit.man,v 1.23 94/04/17 20:43:39 matt Exp $
.\" Copyright (c) 1994  X Consortium
.\" 
.\" Permission is hereby granted, free of charge, to any person obtaining
.\" a copy of this software and associated documentation files (the
.\" "Software"), to deal in the Software without restriction, including
.\" without limitation the rights to use, copy, modify, merge, publish,
.\" distribute, sublicense, and/or sell copies of the Software, and to
.\" permit persons to whom the Software is furnished to do so, subject to
.\" the following conditions:
.\" 
.\" The above copyright notice and this permission notice shall be included
.\" in all copies or substantial portions of the Software.
.\" 
.\" THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
.\" OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
.\" MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
.\" IN NO EVENT SHALL THE X CONSORTIUM BE LIABLE FOR ANY CLAIM, DAMAGES OR
.\" OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
.\" ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
.\" OTHER DEALINGS IN THE SOFTWARE.
.\" 
.\" Except as contained in this notice, the name of the X Consortium shall
.\" not be used in advertising or otherwise to promote the sale, use or
.\" other dealings in this Software without prior written authorization
.\" from the X Consortium.
.\"
.\" $XFree86: xc/programs/xedit/xedit.man,v 1.12 1999/06/20 08:41:42 dawes Exp $
.\"
.TH XEDIT 1 "Release 6" "X Version 11"
.SH NAME
xedit \- simple text editor for X
.SH SYNTAX
.B xedit
[
.I \-toolkitoption
\&.\|.\|. ] [
.I filename
\&.\|.\|. ]
.SH DESCRIPTION
.I Xedit
provides a window consisting of the following four areas:
.IP "Commands Section" 25
A set of commands that allow you to exit \fIxedit\fP, save the file, or
load a new file into the edit window.
.IP "Message Window" 25
Displays \fIxedit\fP messages. In addition, this window can be also used as
a scratch pad. 
.IP "Filename Display"
Displays the name of the file currently being edited, and whether this file
is \fIRead-Write\fP or \fIRead Only\fP.
.IP "Edit Window" 25
Displays the text of the file that you are editing or creating.  
.SH OPTIONS
.I Xedit
accepts all of the standard X Toolkit command line
options (see
.IR X (1)).
The order of the command line options is not important.
.TP 8
.I filename
Specifies the file(s) that are to be loaded during start-up. This is the
file which will be edited. If a file is not specified,
.I xedit
lets you load files or create new files after it has started up.
.SH EDITING
The Athena Text widget is used for the three sections of this
application that allow text input.  The characters typed will go to
the Text widget that has the input focus, or the Text widget that the pointer
cursor is currently over.
.PP

The following keystroke combinations are defined:
.sp
.nf
.ta 1.0i 3.0i 4.5i
Ctrl-a	Beginning Of Line	Meta-b	Backward Word
Ctrl-b	Backward Character	Meta-f	Forward Word
Ctrl-d	Delete Next Character	Meta-i	Insert File
Ctrl-e	End Of Line	Meta-k	Kill To End Of Paragraph
Ctrl-f	Forward Character	Meta-q	Form Paragraph
Ctrl-g	Keyboard Reset	Meta-v	Previous Page
Ctrl-h	Delete Previous Character	Meta-y	Insert Current Selection
Ctrl-j	Newline And Indent	Meta-z	Scroll One Line Down
Ctrl-k	Kill To End Of Line	Meta-d	Delete Next Word
Ctrl-l	Redraw Display	Meta-D	Kill Word
Ctrl-m	Newline	Meta-h	Delete Previous Word
Ctrl-n	Next Line	Meta-H	Backward Kill Word
Ctrl-o	Newline And Backup	Meta-<	Beginning Of File
Ctrl-p	Previous Line	Meta->	End Of File
Ctrl-r	Search/Replace Backward	Meta-]	Forward Paragraph
Ctrl-s	Search/Replace Forward	Meta-[	Backward Paragraph
Ctrl-t	Transpose Characters
Ctrl-u [\fInumber\fP]	Multiply by 4 or \fInumber\fP	Meta-Delete	Delete Previous Word
Ctrl-v	Next Page	Meta-Shift Delete	Kill Previous Word
Ctrl-w	Kill Selection	Meta-Backspace	Delete Previous Word
Ctrl-y	Unkill	Meta-Shift Backspace	Kill Previous Word
Ctrl-z	Scroll One Line Up	Meta-z	Scroll One Line Down
Ctrl-_	Undo
.sp
In addition, the pointer may be used to cut and paste text:
.ta .5i 2.0i
	Button 1 Down	Start Selection
	Button 1 Motion	Adjust Selection
	Button 1 Up	End Selection (cut)

	Button 2 Down	Insert Current Selection (paste)

	Button 3 Down	Extend Current Selection
	Button 3 Motion	Adjust Selection
	Button 3 Up	End Selection (cut)
.fi
.sp
.SH COMMANDS
.IP Quit 8
Quits the current editing session. If any changes have not been saved,
.I xedit 
displays a warning message, allowing the user to save them.
.IP Save 
If file backups are enabled (see RESOURCES, below)
.I xedit
stores a copy of the
original, unedited file in <prefix>\fIfile\fP<suffix>,
then overwrites the \fIfile\fP with the contents of the edit window.  The
filename is retrieved from the Text widget directly to the right of
the \fILoad\fP button.
.IP Load
Loads the file named in the text widget immediately to the right
of the this button and displays it in the Edit window.
.SH RESOURCES
For \fIxedit\fP the available resources are:
.TP 8
.B enableBackups (\fPClass\fB EnableBackups)
Specifies that, when edits made to an existing file are saved,
.I xedit
is to copy the original version of that file to <prefix>\fIfile\fP<suffix>
before it saves the changes.  The default value for this resource is
``on,'' stating that backups should be created.
.TP 8
.B backupNamePrefix (\fPClass\fB BackupNamePrefix)
Specifies a string that is to be prepended to the backup filename.  The
default is that no string shall be prepended.
.TP 8
.B backupNameSuffix (\fPClass\fB BackupNameSuffix)
Specifies a string that is to be appended to the backup filename.  The
default is to use ``~'' as the suffix.
.TP 8
.B positionFormat (\fPClass\fB Format)
Specifies a format string used to display the cursor position.  This
string uses printf(3) like notation, where
.B %l
prints the line number,
.B %c
prints the column number,
.B %p
prints the insert position offset, and
.B %s
prints the current file size.  It is also allowed to specifie field sizes,
with the notation
.B %\-?[0\-9]+
\&.  The default format string is ``L%l'', which shows the character ``L''
followed by the line number.
.TP 8
.B hints (\fPClass\fB Hints)
Specifies a list of strings, separated by new lines, that will be
displayed in the bc_label window.
.TP 8
.B hintsInterval (\fPClass \fBInterval)
Specifies the interval in seconds, which the hint string in the bc_label
window will be changed.
.TP 8
.B changedBitmap (\fPClass \fBBitmap)
Specifies the name of the Bitmap that will be displayed in the fileMenu, when
the file being edited is changed.
.TP 8
.B autoReplace (\fPClass \fBReplace)
This resource is useful to automatically correct common mispelling errors, but
can also be used to create simple macros. The format is
\fI{non-blanks}{blanks}[{string}]\fP. Fields are separeted by newlines.
Example of use:
\"
.br
.nf
nto	  not\\n\\
/macro some long string with \\\\\\n newlines \\\\\\n
.fi
\"
.br
Will automatically replace the word \fInto\fP by \fInot\fP, and \fI/macro\fP
by \fIsome long string with
 newlines
\fP
when you type that words.
.TP 8
.B ispell.dictionaries (\fPClass \fBispell.Dictionary)
Specifies a list of dictionary names, separeted by spaces, available to the
ispell program. The default value is \fI"american americamed+ english"\fP.
.TP 8
.B ispell.dictionary (\fPClass \fBispell.Dictionary)
Specifies the default dictionary to use.
.TP 8
.B ispell*<DICTIONARY>.wordChars (\fPClass \fBispell*Chars)
Specifies a set of characters that can be part of a legal word. The
\fI<DICTIONARY>\fP field is one of the dictionaries specified in the
\fIdictionaries\fP resource.
.TP 8
.B ispell.ispellCommand (\fPClass \fBispell.CommandLine)
The path to the ispell program, and possibly, additional arguments. You don't
need to specify the ``-w'' option, neither the ``-a'' option.
Refer to the \fIispell\fP(1) manpage for more information on ispell options.
.TP 8
.B ispell.formatting (\fPClass \fPispell.TextFormat)
Specifies which text formatting to use while spell checking the file. The
available formats are \fItext\fP and \fIhtml\fP.
.TP 8
.B ispell*text.skipLines (\fPClass \fBispell*text.Skip)
Lines starting with one of the characters in this string will not be spell
checked. This resource is only used in \fItext\fP mode.
.TP 8
.B ispell.terseMode (\fPClass \fBispell.Terse)
When enabled, runs ispell in terse mode, not asking user interaction for words
generated through compound formation (when using the ispell ``-C'' option), or
words generated through affix removal. The default value is False.
.TP 8
.B ispell.lookCommand (\fPClass \fBispell.CommandLine)
The path to the program to search for alternate words, and possibily,
additional arguments. The default program used is \fI/usr/bin/egrep\fP.
.TP 8
.B ispell.wordsFile (\fPClass \fBispell.Words)
The path to the file[s] to search for alternate words. The default file is
\fI/usr/share/dict/words\fP.
.TP 8
.B ispell.guessLabel (\fPClass \fBispell.Status)
String displayed in the ispell status bar when ispell returns a guess list
of one or more words. The default value is \fIGuess\fP.
.TP 8
.B ispell.missLabel (\fPClass \fBispell.Status)
String displayed in the ispell status bar when ispell returns a list of one
or more words to match a mispelled one. The default value is \fIMiss\fP.
.TP 8
.B ispell.rootLabel (\fPClass \fBispell.Status)
String displayed in the ispell status bar when the word is not in the dictionary,
but it can be formed through a root one. The default value is \fIRoot:\fP, and is
followed by a space and the root word.
.TP 8
.B ispell.noneLabel (\fPClass \fBispell.Status)
String displayed in the ispell status bar when there is no near misses. The default
value is \fINone\fP.
.TP 8
.B ispell.compoundLabel (\fPClass \fBispell.Status)
String displayed in the ispell status bar when the word being checked is formed by
concatenation of two words. The default value is \fICompound\fP.
.TP 8
.B ispell.okLabel (\fPClass \fBispell.Status)
String displayed in the ispell status bar when the checked word is in the dictionary.
This string is only displayed when using the \fIcheck\fP button in the xedit ispell
interface. The default value is \fIOk\fP.
.TP 8
.B ispell.eofLabel (\fPClass \fBispell.Status)
The string displayed in the ispell status bar when the end of the file is reached.
The default value is \fIEnd Of File\fP.
.TP 8
.B ispell.repeatLabel (\fPClass \fBispell.Status)
The string displayed in the ispell status bar when two indentical words are found
together in the file. The default value is \fIRepeat\fP.
.TP 8
.B ispell.lookLabel (\fPClass \fBispell.Status)
The string displayed in the ispell status bar after displaying the results of
the \fILook\fP command. If no results are found, the value of the
\fIispell.noneLabel\fP resource is shown.
.TP 8
.B ispell.workingLabel (\fPClass \fBispell.Status)
The string displayed in the ispell status bar while xedit is communicating with
ispell. The default value is \fI...\fP.
.SH WIDGETS
In order to specify resources, it is useful to know the hierarchy of
the widgets which compose \fIxedit\fR.  In the notation below,
indentation indicates hierarchical structure.  The widget class name
is given first, followed by the widget instance name.
.sp
.nf
.ta .5i 1.0i 1.5i 2.0i
Xedit  xedit
	Paned  paned
		Paned  buttons
			Command  quit
			Command  save
			Command  load
			Text     filename
		Label  bc_label
		Text   messageWindow
		Label  labelWindow
		Text   editWindow
.fi
.sp
.SH ENVIRONMENT
.PP
.TP 20
.I DISPLAY
to get the default host and display number.
.TP 20
.I XENVIRONMENT
to get the name of a resource file that overrides the global resources
stored in the RESOURCE_MANAGER property.
.SH FILES
.TP 20
.I <XRoot>/lib/X11/app-defaults/Xedit
specifies required resources
.SH "SEE ALSO"
.IR X (1),
.IR xrdb (1),
.I "Athena Widget Set"
.SH RESTRICTIONS
Xedit is not a replacement to Emacs.
.SH COPYRIGHT
Copyright 1988, Digital Equipment Corporation.
.br
Copyright 1989, X Consortium
.br
Copyright 1998, The XFree86 Project
.br
See
.IR X (1)
for a full statement of rights and permissions.
.SH AUTHORS
Chris D. Peterson, MIT X Consortium
.br
Paulo Cesar Pereira de Andrade, The XFree86 Project


xfree/xc/programs/xedit/xedit.c
/* $XConsortium: xedit.c,v 1.28 94/03/26 17:06:28 rws Exp $ */
 
/*
 *			  COPYRIGHT 1987
 *		   DIGITAL EQUIPMENT CORPORATION
 *		       MAYNARD, MASSACHUSETTS
 *			ALL RIGHTS RESERVED.
 *
 * THE INFORMATION IN THIS SOFTWARE IS SUBJECT TO CHANGE WITHOUT NOTICE AND
 * SHOULD NOT BE CONSTRUED AS A COMMITMENT BY DIGITAL EQUIPMENT CORPORATION.
 * DIGITAL MAKES NO REPRESENTATIONS ABOUT THE SUITABILITY OF THIS SOFTWARE FOR
 * ANY PURPOSE.  IT IS SUPPLIED "AS IS" WITHOUT EXPRESS OR IMPLIED WARRANTY.
 *
 * IF THE SOFTWARE IS MODIFIED IN A MANNER CREATING DERIVATIVE COPYRIGHT RIGHTS,
 * APPROPRIATE LEGENDS MAY BE PLACED ON THE DERIVATIVE WORK IN ADDITION TO THAT
 * SET FORTH ABOVE.
 *
 *
 * Permission to use, copy, modify, and distribute this software and its
 * documentation for any purpose and without fee is hereby granted, provided
 * that the above copyright notice appear in all copies and that both that
 * copyright notice and this permission notice appear in supporting
 * documentation, and that the name of Digital Equipment Corporation not be 
 * used in advertising or publicity pertaining to distribution of the software
 * without specific, written prior permission.
 */
/* $XFree86: xc/programs/xedit/xedit.c,v 1.9 1999/05/23 06:33:53 dawes Exp $ */

#include <X11/IntrinsicP.h>
#include "xedit.h"
#include <X11/Xaw/SmeBSB.h>
#include <time.h>
#include <sys/stat.h>
#include <X11/CoreP.h>

#ifdef X_NOT_STDC_ENV
void srand(); 
int rand(); 
#else
#include <stdlib.h> 
#endif

#define randomize()	srand((unsigned)time((time_t*)NULL))

static XtActionsRec actions[] = {
{"quit", QuitAction},
{"save-file", SaveFile},
{"load-file", LoadFile},
{"find-file", FindFile},
{"cancel-find-file", CancelFindFile},
{"file-completion", FileCompletion},
{"popup-menu", PopupMenu},
{"kill-file", KillFile},
{"split-window", SplitWindow},
{"dir-window", DirWindow},
{"delete-window", DeleteWindow},
{"xedit-focus", XeditFocus},
{"other-window", OtherWindow},
{"switch-source", SwitchSource},
{"ispell", IspellAction},
};

#define DEF_HINT_INTERVAL	300	/* in seconds, 5 minutes */

static Atom wm_delete_window;
static Widget hintswindow;
static int position_format_mask;
static XawTextPositionInfo infos[3];

Widget topwindow, textwindow, messwidget, labelwindow, filenamewindow;
Widget scratch, hpane, vpanes[2], labels[3], texts[3], forms[3], positions[3];
Widget options_popup, dirlabel, dirwindow;
Boolean international;

extern void ResetSourceChanged(xedit_flist_item*);

static void makeButtonsAndBoxes(Widget);
static void HintsTimer(XtPointer, XtIntervalId*);
static void PositionChanged(Widget, XtPointer, XtPointer);
static void StartFormatPosition(void);
static void StartHints(void);

Display *CurDpy;

struct _app_resources app_resources;
struct _xedit_flist flist;

#define Offset(field) XtOffsetOf(struct _app_resources, field)

static XtResource resources[] = {
   {"enableBackups", "EnableBackups", XtRBoolean, sizeof(Boolean),
         Offset(enableBackups), XtRImmediate, FALSE},
   {"backupNamePrefix", "BackupNamePrefix", XtRString, sizeof(char *),
         Offset(backupNamePrefix),XtRString, ""},
   {"backupNameSuffix", "BackupNameSuffix", XtRString, sizeof(char *),
         Offset(backupNameSuffix),XtRString, ".BAK"},
   {"hints", "Hint", XtRString, sizeof(char *),
	 Offset(hints.resource), XtRImmediate, NULL},
   {"hintsInterval", XtCInterval, XtRInt, sizeof(long),
	 Offset(hints.interval), XtRImmediate, (XtPointer)DEF_HINT_INTERVAL},
   {"changedBitmap", XtRBitmap, XtRString, sizeof(char*),
	 Offset(changed_pixmap_name), XtRString, "dot"},
   {"positionFormat", "Format", XtRString, sizeof(char*),
	 Offset(position_format), XtRString, "L%l"},
   {"autoReplace", "Replace", XtRString, sizeof(char*),
	 Offset(auto_replace), XtRImmediate, NULL},
};

#undef Offset

int
main(int argc, char *argv[])
{
  XtAppContext appcon;
  unsigned num_loaded = 0;

  XtSetLanguageProc(NULL, NULL, NULL);
  topwindow = XtAppInitialize(&appcon, "Xedit", NULL, 0, &argc, argv, NULL, NULL, 0);

  XtAppAddActions(appcon, actions, XtNumber(actions));
  XtOverrideTranslations
      (topwindow, XtParseTranslationTable ("<Message>WM_PROTOCOLS: quit()"));
  
  XtGetApplicationResources(topwindow, (XtPointer) &app_resources, resources,
			    XtNumber(resources), NULL, 0);

  CurDpy = XtDisplay(topwindow);
  XawSimpleMenuAddGlobalActions(appcon);
  XtRegisterGrabAction(PopupMenu, True, 
		       ButtonPressMask | ButtonReleaseMask,
		       GrabModeAsync, GrabModeAsync);

  makeButtonsAndBoxes(topwindow);

  StartHints();
  StartFormatPosition();
  (void)StartHooks(appcon);
  if (position_format_mask == 0) {
      int i;

      for (i = 0; i < 3; i++)
	  XtRemoveCallback(texts[i], XtNpositionCallback, PositionChanged, NULL);
  }
  XtRealizeWidget(topwindow);

  options_popup = XtCreatePopupShell("optionsMenu", simpleMenuWidgetClass,
				     topwindow, NULL, 0);
  XtRealizeWidget(options_popup);
  XtAddCallback(XtCreateManagedWidget("ispell", smeBSBObjectClass,
				      options_popup, NULL, 0),
		XtNcallback, IspellCallback, NULL);
  CreateEditPopup();
  
  wm_delete_window = XInternAtom(XtDisplay(topwindow), "WM_DELETE_WINDOW",
				 False);
  (void) XSetWMProtocols (XtDisplay(topwindow), XtWindow(topwindow),
			  &wm_delete_window, 1);

  if (argc > 1) {
      Boolean exists;
      xedit_flist_item *item;
      FileAccess file_access;
      char *filename;
      Widget source;
      Arg args[2];
      unsigned i, num_args;
      char buf[BUFSIZ];

      for (i = 1; i < argc; i++) {
	  struct stat st;

	  num_args = 0;
	  filename = ResolveName(argv[i]);
	  if (filename == NULL || FindTextSource(NULL, filename) != NULL)
	      continue;
	  if (stat(filename, &st) == 0 && !S_ISREG(st.st_mode)) {
	      if (S_ISDIR(st.st_mode)) {
		  if (!num_loaded) {
		      char path[BUFSIZ + 1];

		      strncpy(path, filename, sizeof(path) - 2);
		      path[sizeof(path) - 2] = '\0';
		      if (*path) {
			  if (path[strlen(path) - 1] != '/')
			      strcat(path, "/");
		      }
		      else
			  strcpy(path, "./");
		      XtSetArg(args[0], XtNlabel, "");
		      XtSetValues(dirlabel, args, 1);
		      SwitchDirWindow(True);
		      DirWindowCB(dirwindow, path, NULL);
		      ++num_loaded;
		  }
		  continue;
	      }
	  }

	  switch (file_access = CheckFilePermissions(filename, &exists)) {
	  case NO_READ:
	      if (exists)
		  XmuSnprintf(buf, sizeof(buf), "File %s, %s %s", argv[i],
			      "exists, and could not be opened for",
			      "reading.\n");
	      else
		  XmuSnprintf(buf, sizeof(buf), "File %s %s %s %s", argv[i],
			      "does not exist, and",
			      "the directory could not be opened for",
			      "writing.\n");
	      break;
	  case READ_OK:
	      XtSetArg(args[num_args], XtNeditType, XawtextRead); num_args++;
	      XmuSnprintf(buf, sizeof(buf), "File %s opened READ ONLY.\n",
			  argv[i]);
	      break;
	  case WRITE_OK:
	      XtSetArg(args[num_args], XtNeditType, XawtextEdit); num_args++;
	      XmuSnprintf(buf, sizeof(buf), "File %s opened read - write.\n",
			  argv[i]);
	      break;
	  }
	  if (file_access != NO_READ) {
	      int flags;

	      if (exists) {
		  flags = EXISTS_BIT;
		  XtSetArg(args[num_args], XtNstring, filename);num_args++;
	      }
	      else {
		  flags = 0;
		  XtSetArg(args[num_args], XtNstring, NULL);	num_args++;
	      }
	      source = XtVaCreateWidget("textSource", international ?
					multiSrcObjectClass
					: asciiSrcObjectClass, topwindow,
					XtNtype, XawAsciiFile,
					XtNeditType, XawtextEdit,
					NULL, NULL);
	      XtSetValues(source, args, num_args);
	      item = AddTextSource(source, argv[i], filename,
				   flags, file_access);
	      XtAddCallback(item->source, XtNcallback, SourceChanged,
			    (XtPointer)item);
	      if (exists && file_access == WRITE_OK)
		  item->mode = st.st_mode & (S_IRWXU | S_IRWXG | S_IRWXO);
	      if (!num_loaded)
		  SwitchTextSource(item);
	      ++num_loaded;
	      ResetSourceChanged(item);
	  }
	  XeditPrintf(buf);
      }
  }

    if (!flist.pixmap && strlen(app_resources.changed_pixmap_name)) {
	XrmValue from, to;

	from.size = strlen(app_resources.changed_pixmap_name);
	from.addr = app_resources.changed_pixmap_name;
	to.size = sizeof(Pixmap);
	to.addr = (XtPointer)&(flist.pixmap);

	XtConvertAndStore(flist.popup, XtRString, &from, XtRBitmap, &to);
    }

  if (num_loaded == 0)
      XtSetKeyboardFocus(topwindow, filenamewindow);
  else
      XtSetKeyboardFocus(topwindow, textwindow);

  XtAppMainLoop(appcon);
  exit(0);
}

static void
makeButtonsAndBoxes(Widget parent)
{
    Widget outer, b_row, viewport;
    Arg arglist[10];
    Cardinal num_args;
    xedit_flist_item *item;
    static char *labelWindow = "labelWindow", *editWindow = "editWindow";
    static char *formWindow = "formWindow", *positionWindow = "positionWindow";

    outer = XtCreateManagedWidget("paned", panedWidgetClass, parent,
				  NULL, ZERO);
 
    b_row = XtCreateManagedWidget("buttons", panedWidgetClass, outer, NULL, ZERO);
    {
	MakeCommandButton(b_row, "quit", DoQuit);
	MakeCommandButton(b_row, "save", DoSave);
	MakeCommandButton(b_row, "load", DoLoad);
	filenamewindow = MakeStringBox(b_row, "filename", NULL);
    }
    hintswindow = XtCreateManagedWidget("bc_label", labelWidgetClass,
					outer, NULL, ZERO);

    num_args = 0;
    XtSetArg(arglist[num_args], XtNeditType, XawtextEdit);		++num_args;
    messwidget = XtCreateManagedWidget("messageWindow", asciiTextWidgetClass,
				       outer, arglist, num_args);

    num_args = 0;
    XtSetArg(arglist[num_args], XtNorientation, XtorientHorizontal);	++num_args;
    hpane = XtCreateManagedWidget("hpane", panedWidgetClass, outer,
				  arglist, num_args);

    num_args = 0;
    XtSetArg(arglist[num_args], XtNorientation, XtorientVertical);	++num_args;
    vpanes[0] = XtCreateManagedWidget("vpane", panedWidgetClass, hpane,
				      arglist, num_args);
    XtSetArg(arglist[num_args], XtNheight, 1);				++num_args;
    XtSetArg(arglist[num_args], XtNwidth, 1);				++num_args;
    vpanes[1] = XtCreateWidget("vpane", panedWidgetClass, hpane,
			       arglist, num_args);

    forms[0] = XtCreateManagedWidget(formWindow, formWidgetClass,
				     vpanes[0], NULL, 0);
    labelwindow = XtCreateManagedWidget(labelWindow,labelWidgetClass,
					forms[0], NULL, 0);
    labels[0] = labelwindow;
    positions[0] = XtCreateManagedWidget(positionWindow,labelWidgetClass,
					 forms[0], NULL, 0);

    forms[2] = XtCreateWidget(formWindow, formWidgetClass,
			      vpanes[1], NULL, 0);
    labels[2] = XtCreateManagedWidget(labelWindow,labelWidgetClass,
				      forms[2], NULL, 0);
    positions[2] = XtCreateManagedWidget(positionWindow,labelWidgetClass,
					 forms[2], NULL, 0);

    num_args = 0;
    XtSetArg(arglist[num_args], XtNtype, XawAsciiFile);			++num_args;
    XtSetArg(arglist[num_args], XtNeditType, XawtextEdit);		++num_args;
    textwindow =  XtCreateManagedWidget(editWindow, asciiTextWidgetClass,
					vpanes[0], arglist, num_args);
    num_args = 0;
    XtSetArg(arglist[num_args], XtNinternational, &international);	++num_args;
    XtGetValues(textwindow, arglist, num_args);

    num_args = 0;
    XtSetArg(arglist[num_args], XtNtype, XawAsciiFile);			++num_args;
    XtSetArg(arglist[num_args], XtNeditType, XawtextEdit);		++num_args;
    scratch = XtVaCreateWidget("textSource", international ?
			       multiSrcObjectClass
			       : asciiSrcObjectClass, topwindow,
			       XtNtype, XawAsciiFile,
			       XtNeditType, XawtextEdit,
			       NULL, NULL);
    XtSetValues(scratch, arglist, num_args);

    num_args = 0;
    XtSetArg(arglist[num_args], XtNtextSource, scratch);		++num_args;
    XtSetValues(textwindow, arglist, num_args);

    texts[0] = textwindow;
    num_args = 0;
    XtSetArg(arglist[num_args], XtNtextSource, scratch);		++num_args;
    XtSetArg(arglist[num_args], XtNdisplayCaret, False);		++num_args;
    texts[2] = XtCreateWidget(editWindow, asciiTextWidgetClass,
			      vpanes[1], arglist, num_args);

    forms[1] = XtCreateWidget(formWindow, formWidgetClass,
			      vpanes[0], NULL, 0);
    labels[1] = XtCreateManagedWidget(labelWindow,labelWidgetClass,
				      forms[1], NULL, 0);
    positions[1] = XtCreateManagedWidget(positionWindow,labelWidgetClass,
					 forms[1], NULL, 0);

    texts[1] = XtCreateWidget(editWindow, asciiTextWidgetClass,
			      vpanes[0], arglist, num_args);

    dirlabel = XtCreateWidget("dirlabel", labelWidgetClass,
			      vpanes[1], NULL, 0);
    num_args = 0;
    XtSetArg(arglist[num_args], XtNheight, 1);				++num_args;
    XtSetArg(arglist[num_args], XtNwidth, 1);				++num_args;
    viewport = XtCreateWidget("viewport", viewportWidgetClass,
			      vpanes[1], arglist, num_args);
    dirwindow = XtCreateManagedWidget("dirwindow", listWidgetClass,
				      viewport, NULL, 0);

    item = AddTextSource(scratch, "*scratch*", "*scratch*",
			 0, WRITE_OK);
    XtAddCallback(item->source, XtNcallback, SourceChanged,
		  (XtPointer)item);
    ResetSourceChanged(item);

    for (num_args = 0; num_args < 3; num_args++)
	XtAddCallback(texts[num_args], XtNpositionCallback, PositionChanged, NULL);

    XtAddCallback(dirwindow, XtNcallback, DirWindowCB, NULL);
}

/*	Function Name: Feep
 *	Description: feeps the bell.
 *	Arguments: none.
 *	Returns: none.
 */

void
Feep(void)
{
  XBell(CurDpy, 0);
}

#define	l_BIT		0x01
#define	c_BIT		0x02
#define	p_BIT		0x04
#define	s_BIT		0x08
#define MAX_FMT_LEN	30

static void
StartFormatPosition(void)
{
    char *fmt = app_resources.position_format;

    if (fmt)
	while (*fmt)
	    if (*fmt++ == '%') {
		int len = 0;

		if (*fmt == '-') {
		    ++fmt;
		    ++len;
		}
		while (*fmt >= '0' && *fmt <= '9') {
		    ++fmt;
		    if (++len >= MAX_FMT_LEN) {
			XtAppWarning(XtWidgetToApplicationContext(topwindow),
				     "Format too large to formatPosition");
			position_format_mask = 0;
			return;
		    }
		}
		switch (*fmt++) {
		    case 'l':	position_format_mask |= l_BIT;	break;
		    case 'c':	position_format_mask |= c_BIT;	break;
		    case 'p':	position_format_mask |= p_BIT;	break;
		    case 's':	position_format_mask |= s_BIT;	break;
		    case '%':	break;
		    default: {
			char msg[256];

			XmuSnprintf(msg, sizeof(msg),
				    "Unknown format \"%%%c\" in positionFormat",
				    fmt[-1]);
			XtAppWarning(XtWidgetToApplicationContext(topwindow),
				     msg);
			position_format_mask = 0;
			return;
		    }
		}
	    }
}

/*ARGSUSED*/
static void
PositionChanged(Widget w, XtPointer client_data, XtPointer call_data)
{
    int idx;
    XawTextPositionInfo *info = (XawTextPositionInfo*)call_data;

    for (idx = 0; idx < 3; idx++)
	if (w == texts[idx])
	    break;
    if (idx > 2)
	return;

    if (((position_format_mask & l_BIT)
	  && infos[idx].line_number != info->line_number)
	|| ((position_format_mask & c_BIT)
	    && infos[idx].column_number != info->column_number)
	|| ((position_format_mask & p_BIT)
	    && infos[idx].insert_position != info->insert_position)
	|| ((position_format_mask & s_BIT)
	    && infos[idx].last_position != info->last_position)
	|| infos[idx].overwrite_mode != info->overwrite_mode) {
	int len = 6;
	Arg args[1];
	char buffer[256], *str = app_resources.position_format;
	char fmt_buf[MAX_FMT_LEN + 2], *fmt;

	memcpy(&infos[idx], info, sizeof(XawTextPositionInfo));
	if (info->overwrite_mode)
	    strcpy(buffer, "Ovrwt ");
	else
	    strcpy(buffer, "      ");
	while (*str) {
	    switch (*str) {
		case '%':
		    fmt = fmt_buf;
		    *fmt++ = *str++;
		    if (*str == '-')
			*fmt++ = *str++;
		    /*CONSTCOND*/
		    while (*str >= '0' && *str <= '9') {
			/* StartPositionFormat() already checked the format
			 * length.
			 */
			*fmt++ = *str++;
		    }
		    *fmt++ = 'd';
		    *fmt = '\0';
		    switch (*str) {
			case 'l':
			    XmuSnprintf(&buffer[len], sizeof(buffer) - len,
					fmt_buf, info->line_number);
			    break;
			case 'c':
			    XmuSnprintf(&buffer[len], sizeof(buffer) - len,
					fmt_buf, info->column_number);
			    break;
			case 'p':
			    XmuSnprintf(&buffer[len], sizeof(buffer) - len,
					fmt_buf, info->insert_position);
			    break;
			case 's':
			    XmuSnprintf(&buffer[len], sizeof(buffer) - len,
					fmt_buf, info->last_position);
			    break;
			case '%':
			    strcpy(&buffer[len], "%");
			    break;
		    }
		    len += strlen(&buffer[len]);
		    break;
		default:
		    buffer[len++] = *str;
		    break;
	    }
	    if (len >= sizeof(buffer) - 1)
		break;
	    ++str;
	}
	buffer[len] = '\0';

	XtSetArg(args[0], XtNlabel, buffer);
	XtSetValues(positions[idx], args, 1);
    }
}

/*ARGSUSED*/
static void
HintsTimer(XtPointer closure, XtIntervalId *id)
{
    Arg args[1];
    xedit_hints *hints = (xedit_hints*)closure;

    hints->cur_hint = rand() % hints->num_hints;

    XtSetArg(args[0], XtNlabel, hints->hints[hints->cur_hint]);
    XtSetValues(hintswindow, args, 1);

    hints->timer = XtAppAddTimeOut(XtWidgetToApplicationContext(topwindow),
				   hints->interval, HintsTimer, closure);
}

#define MAX_HINT_LEN		255
#define MIN_HINT_INTERVAL	5
static void
StartHints(void)
{
    char *str, *p;
    unsigned i, len;
    xedit_hints *hints = &(app_resources.hints);

    /* if resource was not set, or was overriden */
    if (hints->resource == NULL || !*hints->resource)
	return;

    randomize();

    if (hints->interval < MIN_HINT_INTERVAL)
	hints->interval = DEF_HINT_INTERVAL;
    hints->interval *= 1000;
    hints->hints = (char**)XtMalloc(sizeof(char*));
    hints->hints[hints->cur_hint = 0] = p = hints->resource;
    hints->num_hints = 1;

    while ((p = strchr(p, '\n')) != NULL) {
	if (*++p == '\0')
	    break;
	hints->hints = (char**)
	    XtRealloc((char*)hints->hints,
		      sizeof(char*) * (hints->num_hints + 1));
	hints->hints[hints->num_hints++] = p;
    }

    /* make a private copy of the resource values, so that one can change
     * the Xrm database safely.
     */
    for (i = 0; i < hints->num_hints; i++) {
	if ((p = strchr(hints->hints[i], '\n')) != NULL)
	    len = p - hints->hints[i];
	else
	    len = strlen(hints->hints[i]);
	if (len > MAX_HINT_LEN)
	    len = MAX_HINT_LEN;
	str = XtMalloc(len + 1);
	strncpy(str, hints->hints[i], len);
	str[len] = '\0';
	hints->hints[i] = str;
    }

    hints->timer = XtAppAddTimeOut(XtWidgetToApplicationContext(topwindow),
				   hints->interval, HintsTimer,
				   (XtPointer)hints);
}
